### PR TITLE
Servlets: Tasks remade with Guice

### PR DESCRIPTION
**Pull request by p-nedelchev**
Servlets task remade with Guice.

### CODE REVIEW

**Comment by p-nedelchev** [comment]
@mgenov PR needs a review.

File: src/main/java/ApplicationServletModule.java [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+        bind(PageViewServlet.class);
+
+        serve("/chain","/first", "/second", "/third").with(Chainlet.class);
+        serve("/counter").with(Counter.class);
+        serve("/greet").with(MainPageServlet.class);
+        serve("/page").with(PageControllerServlet.class);
+        serve("/pageView").with(PageViewServlet.class);
+
+        bind(Integer.class).annotatedWith(Names.named("Port")).toInstance(8080);
+        bind(new TypeLiteral<Map<String, String>>() {
```
Comment (mgenov) [inline src/main/java/ApplicationServletModule.java]: The primitive obsession is forcing you to use annotations when you can just wrap it into an object. PageTemaplates, PageCatalog or something similar.

File: src/main/java/ApplicationServletModule.java [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+        bind(PageControllerServlet.class);
+        bind(PageViewServlet.class);
+
+        serve("/chain","/first", "/second", "/third").with(Chainlet.class);
+        serve("/counter").with(Counter.class);
+        serve("/greet").with(MainPageServlet.class);
+        serve("/page").with(PageControllerServlet.class);
+        serve("/pageView").with(PageViewServlet.class);
+
+        bind(Integer.class).annotatedWith(Names.named("Port")).toInstance(8080);
```
Comment (mgenov) [inline src/main/java/ApplicationServletModule.java]: Port number could be passed as argument from command line.

File: src/main/java/Demo.java [hunk 3]
```diff
@@ -1,9 +1,13 @@
...
+
+
 /**
  * @author Petar Nedelchev (peter.krasimirov@gmail.com)
  */
 public class Demo {
     public static void main(String[] args) {
-        Jetty jetty = new Jetty(8080);
-        jetty.start();
+        Guice.createInjector(new ApplicationServletModule()).getInstance(Jetty.class).start();
```
Comment (mgenov) [inline src/main/java/Demo.java]: You have to stop your service (the HTTP service) gracefully when app is going to shutdown. 

Take a look at: 

```java
Runtime.getRuntime().addShutdownHook(...)
```

Cleanups that need to be perfomed could be executed in the shutdown hook.

File: src/main/java/greeting/MainPageServlet.java [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+import javax.servlet.http.HttpServletResponse;
+import java.io.IOException;
+
+/**
+ * @author Petar Nedelchev (peter.krasimirov@gmail.com)
+ */
+@Singleton
+public class MainPageServlet extends HttpServlet {
+
+    public MainPageServlet() {
```
Comment (mgenov) [inline src/main/java/greeting/MainPageServlet.java]: this constructor is not required.

File: src/main/java/ApplicationServletModule.java (hunk 2 above)
Comment (p-nedelchev) [inline src/main/java/ApplicationServletModule.java]: Port can be passed as argument now.

File: src/main/java/ApplicationServletModule.java (hunk 1 above)
Comment (p-nedelchev) [inline src/main/java/ApplicationServletModule.java]: Wrapped collections in objects.

File: src/main/java/greeting/MainPageServlet.java (hunk 4 above)
Comment (p-nedelchev) [inline src/main/java/greeting/MainPageServlet.java]: Removed.

File: src/main/java/Demo.java (hunk 3 above)
Comment (p-nedelchev) [inline src/main/java/Demo.java]: used `public void setGracefulShutdown(int timeoutMS)` 
ref: http://download.eclipse.org/jetty/8.1.17.v20150415/apidocs/org/eclipse/jetty/server/Server.html#getGracefulShutdown()

File: src/main/java/Demo.java [hunk 5]
```diff
@@ -1,9 +1,28 @@
...
     public static void main(String[] args) {
-        Jetty jetty = new Jetty(8080);
+        Integer port = Ints.tryParse(args[0]);
+        if(port == null) {
+            Logger.getLogger("DefaultPort").log(Level.INFO, "Server started at default port: 8080");
+            port = 8080;
+        }
+        Injector injector = Guice.createInjector(new ApplicationServletModule(port));
+        Jetty jetty = injector.getInstance(Jetty.class);
         jetty.start();
```
Comment (mgenov) [inline src/main/java/Demo.java]: Are you sure that jetty.start() is not blocking? If it's blocking then the runtime code will not be executed.

Comment (p-nedelchev) [inline src/main/java/Demo.java]: I'm not really sure. I better register the shutdownHook before jetty start method.

**Comment by p-nedelchev** [comment]
PR is updated.

**Comment by mgenov** [comment]
PR is not up-to date. 

screen shot 2017-02-13 at 6 40 15 pm

**Pull request closed without merging**